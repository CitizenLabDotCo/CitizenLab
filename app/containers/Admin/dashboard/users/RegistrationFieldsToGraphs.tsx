// libraries
import React, { PureComponent } from 'react';
import { map } from 'lodash-es';

// resources
import GetUserCustomFields, {
  GetUserCustomFieldsChildProps,
} from 'modules/user_custom_fields/resources/GetUserCustomFields';
import { isNilOrError } from 'utils/helperUtils';

// services
import {
  usersByRegFieldStream,
  IUsersByRegistrationField,
  usersByRegFieldXlsxEndpoint,
} from 'services/stats';

// intl
import { injectIntl } from 'utils/cl-intl';
import { InjectedIntlProps } from 'react-intl';
import localize, { InjectedLocalized } from 'utils/localize';
import messages from '../messages';

// components
import BarChartByCategory from './charts/BarChartByCategory';
import PieChartByCategory from './charts/PieChartByCategory';
import AreaChart from './charts/AreaChart';
import GenderChart from './charts/GenderChart';
import AgeChart from './charts/AgeChart';

interface InputProps {
  currentGroupFilter: string | undefined;
  currentGroupFilterLabel: string | undefined;
  startAt: string | null | undefined;
  endAt: string | null;
}

interface DataProps {
  customFields: GetUserCustomFieldsChildProps;
}

export interface Props extends InputProps, DataProps {}

export class RegistrationFieldsToGraphs extends PureComponent<
  Props & InjectedIntlProps & InjectedLocalized
> {
  convertToGraphFormat = (data: IUsersByRegistrationField) => {
    const {
      series: { users },
      options,
    } = data;
    const res = map(options, (value, key) => {
      return {
        value: users[key] || 0,
        name: this.props.localize(value.title_multiloc),
        code: key,
      };
    });

    if (users['_blank']) {
      res.push({
        value: users['_blank'],
        name: this.props.intl.formatMessage(messages._blank),
        code: '_blank',
      });
    }

    return res.length > 0 ? res : null;
  };
  convertCheckboxToGraphFormat = (data: IUsersByRegistrationField) => {
    const {
      series: { users },
    } = data;
    const res = ['_blank', 'true', 'false'].map((key) => ({
      value: users[key] || 0,
      name: this.props.intl.formatMessage(messages[key]),
      code: 'key',
    }));

    return res.length > 0 ? res : null;
  };

  render() {
    const {
      customFields,
      localize,
      startAt,
      endAt,
      currentGroupFilter,
      currentGroupFilterLabel,
    } = this.props;

    if (isNilOrError(customFields)) {
      return null;
    }

    return customFields.map((field, index) => {
      if (field.attributes.enabled) {
        if (field.attributes.code === 'birthyear') {
          return (
            <AgeChart
              key={index}
              startAt={startAt}
              endAt={endAt}
              currentGroupFilter={currentGroupFilter}
              currentGroupFilterLabel={currentGroupFilterLabel}
            />
          );
        }
        if (field.attributes.input_type === 'number') {
          return;
        }
        if (field.attributes.code === 'gender') {
          return (
            <GenderChart
              key={index}
              startAt={startAt}
              endAt={endAt}
              currentGroupFilter={currentGroupFilter}
              currentGroupFilterLabel={currentGroupFilterLabel}
            />
          );
        }

        if (field.attributes.code === 'domicile') {
          return (
            <AreaChart
              key={index}
              startAt={startAt}
              endAt={endAt}
              currentGroupFilter={currentGroupFilter}
              currentGroupFilterLabel={currentGroupFilterLabel}
            />
          );
        }

        if (field.attributes.input_type === 'checkbox') {
          return (
            <PieChartByCategory
              key={index}
              startAt={startAt}
              endAt={endAt}
              currentGroupFilter={currentGroupFilter}
              currentGroupFilterLabel={currentGroupFilterLabel}
              convertToGraphFormat={this.convertCheckboxToGraphFormat}
              graphTitleString={localize(field.attributes.title_multiloc)}
              stream={usersByRegFieldStream}
              graphUnit="users"
              customId={field.id}
              xlsxEndpoint={usersByRegFieldXlsxEndpoint(field.id)}
            />
          );
        } else {
          return (
            <BarChartByCategory
              key={index}
              startAt={startAt}
              endAt={endAt}
              currentGroupFilter={currentGroupFilter}
              currentGroupFilterLabel={currentGroupFilterLabel}
              convertToGraphFormat={this.convertToGraphFormat}
              graphTitleString={localize(field.attributes.title_multiloc)}
              stream={usersByRegFieldStream}
              graphUnit="users"
              customId={field.id}
              xlsxEndpoint={usersByRegFieldXlsxEndpoint(field.id)}
            />
          );
        }
      }

      return null;
    });
  }
}

const RegistrationFieldsToGraphsWithHoCs = localize<Props>(
  injectIntl<Props & InjectedLocalized>(RegistrationFieldsToGraphs as any)
) as any;

export default (inputProps: InputProps) => (
  <GetUserCustomFields
    inputTypes={['select', 'multiselect', 'checkbox', 'number']}
  >
    {(customFields) => (
      <RegistrationFieldsToGraphsWithHoCs
        {...inputProps}
        customFields={customFields}
      />
    )}
  </GetUserCustomFields>
);
