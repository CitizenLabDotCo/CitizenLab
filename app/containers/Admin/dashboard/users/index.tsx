// libraries
import React, { PureComponent } from 'react';
import moment from 'moment';
import { ThemeProvider } from 'styled-components';

// components
import {
  chartTheme,
  Container,
  GraphsContainer,
  Line,
  GraphCard,
  GraphCardInner,
  GraphCardTitle,
  ControlBar
} from '../';
import TimeControl from '../components/TimeControl';
import IntervalControl from '../components/IntervalControl';
import GenderChart from '../components/GenderChart';
import AgeChart from '../components/AgeChart';

// i18n
import { FormattedMessage } from 'utils/cl-intl';
import messages from '../messages';

interface State {
  interval: 'weeks' | 'months' | 'years';
  intervalIndex: number;
  currentGroupFilter?: string;
}

type Props = {

};

export default class UsersDashboard extends PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      interval: 'months',
      intervalIndex: 0,
      currentGroupFilter: undefined
    };
  }

  changeInterval = (interval: 'weeks' | 'months' | 'years') => {
    this.setState({ interval, intervalIndex: 0 });
  }

  changeIntervalIndex = (intervalIndex: number) => {
    this.setState({ intervalIndex });
  }

  handleOnGroupFilter = (filter) => {
    // To be implemented
    this.setState({ currentGroupFilter: filter });
  }

  render() {
    const { interval, intervalIndex, currentGroupFilter } = this.state;
    const startAtMoment = moment().startOf(interval).add(intervalIndex, interval);
    const endAtMoment = moment(startAtMoment).add(1, interval);
    const startAt = startAtMoment.toISOString();
    const endAt = endAtMoment.toISOString();

    return (
      <Container>
        <ControlBar>
          <TimeControl
            value={intervalIndex}
            interval={interval}
            onChange={this.changeIntervalIndex}
            currentTime={startAtMoment}
          />
          <IntervalControl
            value={interval}
            onChange={this.changeInterval}
          />
        </ControlBar>
        <ThemeProvider theme={chartTheme}>
          <GraphsContainer>
            <Line>
              <GraphCard className="first halfWidth">
                <GraphCardInner>
                  <GraphCardTitle>
                    <FormattedMessage {...messages.usersByGenderTitle} />
                  </GraphCardTitle>
                  <GenderChart startAt={startAt} endAt={endAt} currentGroupFilter={currentGroupFilter} />
                </GraphCardInner>
              </GraphCard>
              <GraphCard className="halfWidth">
                <GraphCardInner>
                  <GraphCardTitle>
                    <FormattedMessage {...messages.usersByAgeTitle} />
                  </GraphCardTitle>
                  <AgeChart startAt={startAt} endAt={endAt} currentGroupFilter={currentGroupFilter} />
                </GraphCardInner>
              </GraphCard>
            </Line>
          </GraphsContainer>
        </ThemeProvider>
      </Container>
    );
  }
}
