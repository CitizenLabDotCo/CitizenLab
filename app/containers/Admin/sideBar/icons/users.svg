<svg xmlns="http://www.w3.org/2000/svg" width="21" height="22" viewBox="0 0 21 22">
    <g fill="none" fillRule="evenodd" stroke="#FFF" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" opacity=".6" transform="translate(1 1)">
        <circle cx="9.091" cy="10" r="9.091"/>
        <circle cx="9.091" cy="8.977" r="2.537"/>
        <path d="M9.09 12.198a5.074 5.074 0 0 0-5.065 5.346 9.09 9.09 0 0 0 5.066 1.547 9.09 9.09 0 0 0 5.06-1.548c.007-.09.012-.18.014-.27a5.074 5.074 0 0 0-5.074-5.075z"/>
    </g>
</svg>
