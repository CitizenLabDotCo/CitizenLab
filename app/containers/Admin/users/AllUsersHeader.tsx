// Libraries
import React from 'react';
import { throttle } from 'lodash';

// Styling
import { TitleWrapper, TextAndButtons, StyledSearch, Spacer } from './GroupHeader';

// i18l
import { FormattedMessage } from 'utils/cl-intl';
import messages from './messages';

// Typings
export interface Props {
  onSearch: (newValue: string) => void;
}
export interface State {
  searchValue: string;
}

export class AllUsersHeader extends React.PureComponent<Props, State> {
  debounceSearch: (newValue: string) => void;

  constructor(props) {
    super(props);
    this.state = {
      searchValue: '',
    };
    this.debounceSearch = throttle(this.props.onSearch, 500);
  }

  handleSearchChange = (newValue: string) => {
    this.setState({ searchValue: newValue });
    this.debounceSearch(newValue);
  }

  render() {
    return (
      <TitleWrapper>
        <TextAndButtons>
          <FormattedMessage tagName="h1" {...messages.allUsers} />
        </TextAndButtons>
        <Spacer />
        <StyledSearch value={this.state.searchValue} onChange={this.handleSearchChange} />
      </TitleWrapper>
    );
  }
}

export default AllUsersHeader;
