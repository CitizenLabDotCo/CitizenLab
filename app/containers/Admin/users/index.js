import React from 'react';
import PropTypes from 'prop-types';
import HelmetIntl from 'components/HelmetIntl';

// components
// import { FormattedMessage } from 'react-intl';
// import WatchSagas from 'utils/containers/watchSagas';
// import sagas from 'resources/users/sagas';

// store
// import { preprocess } from 'utils';

// messages
import messages from './messages';

class UserDashboard extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function

  render() {
    return (
      <div>
        <HelmetIntl
          title={messages.helmetTitle}
          description={messages.helmetDescription}
        />
        {this.props.children}
      </div>

    );
  }
}


UserDashboard.propTypes = {
  children: PropTypes.element,
};

export default UserDashboard;
