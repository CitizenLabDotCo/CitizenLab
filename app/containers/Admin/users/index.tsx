import React, { SFC } from 'react';

// components
import HelmetIntl from 'components/HelmetIntl';
import TabbedResource from 'components/admin/TabbedResource';

// resources
import GetLocation, { GetLocationChildProps } from 'resources/GetLocation';

// i18n
import messages from './messages';
import { InjectedIntlProps } from 'react-intl';
import { injectIntl } from 'utils/cl-intl';

interface InputProps {}

interface DataProps {
  location: GetLocationChildProps;
}

interface Props extends InputProps, DataProps {}

const UsersPage: SFC<Props & InjectedIntlProps> = ({ location, children, intl }) => {
  const { formatMessage } = intl;

  const tabs = [
    { label: formatMessage(messages.tabRegisteredUsers), url: '/admin/users/registered' },
    { label: formatMessage(messages.tabInviteByEmail), url: '/admin/users/invitations' },
  ];

  const resource = {
    title: formatMessage(messages.viewPublicResource)
  };

  if (!location) return null;

  return (
    <>
      <TabbedResource
        resource={resource}
        messages={messages}
        tabs={tabs}
        location={location}
      >
        <HelmetIntl
          title={messages.helmetTitle}
          description={messages.helmetDescription}
        />
        {children}
      </TabbedResource>
    </>
  );
};

const UsersPageWithIntl = injectIntl(UsersPage);

export default (inputProps: InputProps) => (
  <GetLocation>
    {location => <UsersPageWithIntl {...inputProps} location={location} />}
  </GetLocation>
);
