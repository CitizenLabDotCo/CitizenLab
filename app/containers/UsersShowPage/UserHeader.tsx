import React, { memo } from 'react';
import { isEmpty } from 'lodash-es';
import { isNilOrError } from 'utils/helperUtils';
import moment from 'moment';

// components
import Avatar from 'components/Avatar';
import QuillEditedContent from 'components/UI/QuillEditedContent';

// resources
import GetWindowSize, { GetWindowSizeChildProps } from 'resources/GetWindowSize';
import GetUser, { GetUserChildProps } from 'resources/GetUser';
import GetAuthUser, { GetAuthUserChildProps } from 'resources/GetAuthUser';

// i18n
import { FormattedMessage } from 'utils/cl-intl';
import T from 'components/T';
import messages from './messages';

// style
import styled from 'styled-components';
import { colors, fontSizes, media, viewportWidths } from 'utils/styleUtils';
import { adopt } from 'react-adopt';
import Link from 'utils/cl-router/Link';
import Icon from 'components/UI/Icon';
import { darken } from 'polished';

const Container = styled.div`
  background-color: white;
  width: 100%;
  padding-top: 30px;
  padding-bottom: 70px;

  ${media.smallerThanMaxTablet`
    padding: 20px 20px 35px;
  `}
`;

const UserAvatar = styled.div`
  width: 100%;
  display: flex;
  justify-content: center;
  margin-bottom: 40px;
`;

const UserInfo = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
  margin-top: 0px;
  font-size: ${fontSizes.base}px;
`;

const FullName = styled.div`
  width: 100%;
  padding-top: 0px;
  font-size: ${fontSizes.xl}px;
  font-weight: 600;
  text-align: center;
  color:  ${({ theme }) => theme.colorText};
`;

const JoinedAt = styled.div`
  width: 100%;
  margin-top: 10px;
  font-weight: 300;
  text-align: center;
  color:  ${({ theme }) => theme.colorText};
`;

const Bio = styled.div`
  line-height: 1.25;
  max-width: 600px;
  text-align: center;
  font-weight: 400;
  margin: 18px auto;
`;

const EditProfile = styled(Link)`
  margin-top: 10px;
  padding: 6px 10px;
  color: ${colors.label};
  border-radius: ${({ theme }) => theme.borderRadius};
  display: flex;
  align-items: center;
  transition: all 100ms ease-out;

  &:hover,
  &:focus {
    color: ${darken(0.2, colors.label)};
    background: ${colors.background};
  }
`;

const EditIcon = styled(Icon)`
  margin-right: 8px;
`;

interface InputProps {
  userSlug: string | null;
}

interface DataProps {
  windowSize: GetWindowSizeChildProps;
  user: GetUserChildProps;
  authUser: GetAuthUserChildProps;
}

interface Props extends InputProps, DataProps {}

export const UserHeader = memo<Props>(props => {
    const { user, authUser, windowSize } = props;
    const smallerThanSmallTablet = windowSize ? windowSize <= viewportWidths.smallTablet : false;

    if (!isNilOrError(user)) {
      const memberSinceMoment = moment(user.attributes.created_at).format('LL');

      return (
        <Container>
          <UserAvatar>
            <Avatar userId={user.id} size={smallerThanSmallTablet ? '120px' : '150px'} />
          </UserAvatar>

          <UserInfo>
            <FullName id="e2e-usersshowpage-fullname">{user.attributes.first_name} {user.attributes.last_name}</FullName>
            <JoinedAt>
              <FormattedMessage {...messages.memberSince} values={{ date: memberSinceMoment }} />
            </JoinedAt>
            {!isEmpty(user.attributes.bio_multiloc) &&
              <Bio>
                <QuillEditedContent>
                  {user.attributes.bio_multiloc && <T value={user.attributes.bio_multiloc} supportHtml={true} />}
                </QuillEditedContent>
              </Bio>
            }
            {!isNilOrError(authUser) && authUser.id === user.id &&
              <EditProfile to="/profile/edit">
                <EditIcon name="pencil" />
                <FormattedMessage {...messages.editProfile} />
              </EditProfile>
            }
          </UserInfo>
        </Container>
      );
    }

    return null;
  }
);

const Data = adopt<DataProps, InputProps>({
  windowSize: <GetWindowSize />,
  authUser: <GetAuthUser />,
  user: ({ userSlug, render }) => <GetUser slug={userSlug}>{render}</GetUser>
});

const WrappedUserHeader = (inputProps: InputProps) => (
  <Data {...inputProps}>
    {dataProps => <UserHeader {...inputProps} {...dataProps} />}
  </Data>
);

export default WrappedUserHeader;
