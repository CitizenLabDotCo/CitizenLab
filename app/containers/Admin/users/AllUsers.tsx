// Libraries
import React from 'react';
import { isString, isEmpty } from 'lodash';

// Components
import UserManager from './UserManager';
import AllUsersHeader from './AllUsersHeader';


interface InputProps { }


interface Props extends InputProps { }

interface State {
  search: string | undefined;
}


export default class AllUsers extends React.PureComponent<Props, State> {

  constructor(props) {
    super(props);
    this.state = {
      search: undefined,
    };
  }

  searchUser = (searchTerm: string) => {
    this.setState({
      search: (isString(searchTerm) && !isEmpty(searchTerm) ? searchTerm : '')
    });
  }

  render() {
    const { search } = this.state;

    console.log(this.state);

    return (
      <>
        <AllUsersHeader onSearch={this.searchUser} />

          <UserManager
            search={search}
          />
      </>
    );
  }
}
