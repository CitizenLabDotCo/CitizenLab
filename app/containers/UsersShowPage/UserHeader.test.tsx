// libraries
import React from 'react';
import { shallow } from 'enzyme';

// component to test
import { UserHeader } from './UserHeader';

// mock utilities
jest.mock('utils/cl-intl');
jest.mock('services/comments');
import { makeUser } from 'services/__mocks__/users';

describe('<UserHeader />', () => {

  it('renders correctly on some user without a bio', () => {
    const user = makeUser();
    const Wrapper = shallow(<UserHeader user={user.data} />);
    expect(Wrapper).toMatchSnapshot();
  });
  it('renders correctly on some user with a bio', () => {
    const user = makeUser({ bio_multiloc: { en: 'Hi there ! ' } });
    const Wrapper = shallow(<UserHeader user={user.data} />);
    expect(Wrapper).toMatchSnapshot();
  });
  it('renders correctly on your own profile', () => {
    const user = makeUser({}, 'someUser');
    const Edit = shallow(<UserHeader user={user.data} authUser={{ id: 'someUser' }} />)
    .find('UserHeader__EditProfile');
    expect(Edit).toMatchSnapshot();
  });
  it('renders correctly when user is null', () => {
    const Edit = shallow(<UserHeader user={null} />)
    .find('UserHeader__EditProfile');
    expect(Edit).toMatchSnapshot();
  });

});
