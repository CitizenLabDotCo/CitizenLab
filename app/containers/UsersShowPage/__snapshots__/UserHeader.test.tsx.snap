// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserHeader /> renders correctly on some user with a bio 1`] = `
<UserHeader__Container>
  <UserHeader__UserAvatar>
    <WrappedAvatar
      size="105px"
      userId="522ae8cc-a5ed-4d31-9aa0-470904934ec6"
    />
  </UserHeader__UserAvatar>
  <UserHeader__UserInfo>
    <UserHeader__FullName
      id="e2e-usersshowpage-fullname"
    >
      Test
       
      Citizenlab
    </UserHeader__FullName>
    <UserHeader__JoinedAt>
      <FormattedMessage
        defaultMessage="Member since {date}"
        id="app.containers.UsersShowPage.memberSince"
        values={
          Object {
            "date": "November 26, 2018",
          }
        }
      />
    </UserHeader__JoinedAt>
    <UserHeader__Bio>
      <WithTheme(QuillEditedContent)>
        <T
          supportHtml={true}
          value={
            Object {
              "en": "Hi there ! ",
            }
          }
        />
      </WithTheme(QuillEditedContent)>
    </UserHeader__Bio>
  </UserHeader__UserInfo>
</UserHeader__Container>
`;

exports[`<UserHeader /> renders correctly on some user without a bio 1`] = `
<UserHeader__Container>
  <UserHeader__UserAvatar>
    <WrappedAvatar
      size="105px"
      userId="522ae8cc-a5ed-4d31-9aa0-470904934ec6"
    />
  </UserHeader__UserAvatar>
  <UserHeader__UserInfo>
    <UserHeader__FullName
      id="e2e-usersshowpage-fullname"
    >
      Test
       
      Citizenlab
    </UserHeader__FullName>
    <UserHeader__JoinedAt>
      <FormattedMessage
        defaultMessage="Member since {date}"
        id="app.containers.UsersShowPage.memberSince"
        values={
          Object {
            "date": "November 26, 2018",
          }
        }
      />
    </UserHeader__JoinedAt>
  </UserHeader__UserInfo>
</UserHeader__Container>
`;

exports[`<UserHeader /> renders correctly on your own profile 1`] = `
<UserHeader__EditProfile
  to="/profile/edit"
>
  <Component
    name="pencil"
  />
  <FormattedMessage
    defaultMessage="Edit my profile"
    id="app.containers.UsersShowPage.editProfile"
  />
</UserHeader__EditProfile>
`;

exports[`<UserHeader /> renders correctly when user is null 1`] = `null`;
